// undici/index-fetch.js
'use strict'

const { getGlobalDispatcher, setGlobalDispatcher } = require('undici/lib/global')
const EnvHttpProxyAgent = require('undici/lib/dispatcher/env-http-proxy-agent')
const fetchImpl = require('undici/lib/web/fetch').fetch

module.exports.fetch = function fetch (init, options = undefined) {
  return fetchImpl(init, options).catch(err => {
    if (err && typeof err === 'object') {
      Error.captureStackTrace(err)
    }
    throw err
  })
}
module.exports.FormData = require('undici/lib/web/fetch/formdata').FormData
module.exports.Headers = require('undici/lib/web/fetch/headers').Headers
module.exports.Response = require('undici/lib/web/fetch/response').Response
module.exports.Request = require('undici/lib/web/fetch/request').Request

const { CloseEvent, ErrorEvent, MessageEvent, createFastMessageEvent } = require('undici/lib/web/websocket/events')
module.exports.WebSocket = require('undici/lib/web/websocket/websocket').WebSocket
module.exports.CloseEvent = CloseEvent
module.exports.ErrorEvent = ErrorEvent
module.exports.MessageEvent = MessageEvent
module.exports.createFastMessageEvent = createFastMessageEvent

module.exports.EventSource = require('undici/lib/web/eventsource/eventsource').EventSource

const api = require('undici/lib/api')
const Dispatcher = require('undici/lib/dispatcher/dispatcher')
Object.assign(Dispatcher.prototype, api)
// Expose the fetch implementation to be enabled in Node.js core via a flag
module.exports.EnvHttpProxyAgent = EnvHttpProxyAgent
module.exports.getGlobalDispatcher = getGlobalDispatcher
module.exports.setGlobalDispatcher = setGlobalDispatcher
